core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -662,7 +660,7 @@ public void close()
             bufferedPageCount = 0;
             writeBuffer = null;
 
-            if (exchangeSink != null && !exchangeSinkFinished.get()) {
+            if (exchangeSink != null) {
                 try {
                     exchangeSink.abort().whenComplete((result, failure) -> {
```

Comment:
can `abort` even happen in practice now since you set `exchangeSink = null` immediately after `exchangeSink.finish()`?

Reply:
To ensure `abort` is called after `finish` is started the `exchangeSink` variable would have to be nullified only after `finish` is done. That would mean that the `exchangeSink` variable has to be set to `null` from a different thread what requires synchronization. 

Adding efficient synchronization to `PageBuffer#exchangeSink` variable will result in increased complexity. The `ExchangeSink#abort` is not guaranteed to abort the `finish` process, and the `finish` operation is required to guarantee proper resource release in case of it's failure. So calling `abort` after `finish` is technically a best effort?

Given a very narrow scope and a low likelihood of this condition I wonder if it makes sense to increase complexity to support it? What do you think?

Reply:
Discussed offline. Added synchronization.